### ISSUE TITLE
UnnecessaryFilter false positive

### ISSUE THREAD

**Issue by vladimirfx** [detekt/detekt#3541]
## Expected Behavior
Code not flagged.

## Observed Behavior
Correct code flagger.

## Steps to Reproduce

``` kotlin
        val  queryParts: List<String> = ...
        val dbQueryParts = queryParts.filter { it.length > 1 }.take(3)
```

Second line reported 3 times:

```
        UnnecessaryFilter - [dbQueryParts] at **/PlaceHandler.kt:41:39
        UnnecessaryFilter - [dbQueryParts] at **/PlaceHandler.kt:41:39
        UnnecessaryFilter - [dbQueryParts] at **/PlaceHandler.kt:41:39
 ```

```
    val listOfPairs: List<Pair<String, String>> = emptyList()
    listOfPairs.filter { (a, b) -> a.equals(b, true) }.map { it.first }
```

Second line reported 1 time.

## Your Environment

* Version of detekt used: 1.16.0
* Version of Gradle used (if applicable): 6.8.3
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: MacOS
* Link to your project (if it's a public repository):

**Comment by BraisGabin** [comment]
🤔 I just created tests with both of your snippets and they work as expected. Could you provide a sample file (or project) so we can reproduce your issue?

**Comment by wilerson** [comment]
Not sure whether it's the same issue, but I'm seeing the following case:

```kotlin
val strings = listOf("abc", "cde", "ader")
val filteredStrings = strings.filter { "a" in it }
filteredStrings.forEach { println(it) }
if (filteredStrings.size > 2) {
    println("more than two")
}
```

**Comment by BraisGabin** [comment]
🤔 this is strange. I can't reproduce it in the tests but if I add the code in our project and run detekt it finds the issue. Anyway I confirm this issue.

**Comment by BraisGabin** [comment]
Ok, to reproduce this issue in the unit tests we need to add it inside a function:

```kotlin
fun foo() {
    val strings = listOf("abc", "cde", "ader")
    val filteredStrings = strings.filter { "a" in it }
    filteredStrings.forEach { println(it) }
    if (filteredStrings.size > 2) {
        println("more than two")
    }
}
```
